src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs

```diff
@@ -347,6 +347,10 @@ public static bool InterlockedInitialize<T>(ref ImmutableArray<T> location, Immu
                 if (priorCollection.TryGetValue(key, out oldValue!))
                 {
                     newValue = updateValueFactory(key, oldValue);
+                    if (priorCollection.ValueComparer.Equals(oldValue, newValue))
```

Comment:
💭 It looks like `SetItem` already performs a value check. If you are looking for a way to ensure this method does not end up performing more [CAS](https://en.wikipedia.org/wiki/Compare-and-swap) operations than necessary, we could do a reference equality check between `updatedCollection` and `priorCollection` between the call to `SetItem` and the call to `CompareExchange`. This also avoids any issues with the use of a specific value comparer and virtual/interface dispatch overhead it could incur.

Reply:
@sharwell Fixed, thanks.